sdks/python/apache_beam/io/range_trackers.py

```diff
@@ -133,29 +133,44 @@ def set_current_position(self, record_start):
       self._validate_record_start(record_start, False)
       self._last_record_start = record_start
 
+  def _validate_split_offset(self, split_offset):
+    if split_offset <= self._last_record_start:
+      logging.debug(
+          'Split position %s at %r is not valid: '
+          'already past proposed stop offset', split_offset, self)
+      return False
+    if (split_offset < self.start_position()
+        or split_offset >= self.stop_position()):
+      logging.debug(
+          'Split position %s at %s is not valid: '
+          'proposed split position out of range', split_offset, self)
+      return False
+    return True
+
   def try_split(self, split_offset):
     assert isinstance(split_offset, (int, long))
     with self._lock:
       if self._stop_offset == OffsetRangeTracker.OFFSET_INFINITY:
-        logging.debug('refusing to split %r at %d: stop position unspecified',
+        logging.debug('Refusing to split %r at %d: stop position unspecified',
                       self, split_offset)
         return
+
       if self._last_record_start == -1:
         logging.debug('Refusing to split %r at %d: unstarted', self,
                       split_offset)
         return
 
-      if split_offset <= self._last_record_start:
-        logging.debug(
-            'Refusing to split %r at %d: already past proposed stop offset',
-            self, split_offset)
-        return
-      if (split_offset < self.start_position()
-          or split_offset >= self.stop_position()):
-        logging.debug(
-            'Refusing to split %r at %d: proposed split position out of range',
-            self, split_offset)
-        return
+      if not self._validate_split_offset(split_offset):
+        split_offset = self._last_attempted_record_start + 1
```

Comment:
I'm bit worried about the change of behavior due to this. This might end up creating lot more splits that we do today. But then again this might be OK since backend already asked the SDK to split. Did you run any large scale tests (for example, that use textio) to determine if you see any difference in performance characteristics ?

Also, what's the motivation for this change ? Do you think performance will improve since we re-split at an earlier position ? If so I think this should be backed by performance benchmarks that verifies that this change actually helps.

Reply:
Also, please add a JIRA.